### ISSUE TITLE
net: Allow setting socket options with DialTCP

### ISSUE THREAD

**Issue by dorianperkins** [golang/go#34743]
Previous scale testing has shown that `net.Dial` is more expensive than `net.DialTCP` (https://github.com/golang/go/issues/18601
) since it performs actions such as re-resolving the address on each execution. Since we're making many connections to the same IP:port, we currently avoid these performance penalties by using `net.DialTCP` which allows us to pass in a pre-resolved `*net.TCPAddr`.

For performance reasons, we also need to set socket options within this same codepath.
As of Go 1.11, `net.Dialer` exposes a `Control` function (https://github.com/golang/go/issues/9661) which allows setting socket options, but this requires using the more expensive `net.Dial` path. 

Today, it looks like we only have 2 options: 
1) set socket options but sacrifice performance by using `net.Dial`, or 
2) we continue to use `net.DialTCP` but cannot set socket options. 

Is the Go team open to exposing some way to use `net.DialTCP` which allows setting socket options? We're happy to contribute code if needed.

**Comment by katiehockman** [comment]
Seems related to https://github.com/golang/go/issues/18601

/cc @bradfitz

**Comment by bradfitz** [comment]
What performance numbers do you see between the two?

Can you summarize where you're seeing the performance loss between Dial and DialTCP? #18601 looks like it's kinda long and about an old Go version and talking about things that are common to both Dial and DialTCP. 

If the main cost is resolving the ip:port from the name, can't you do that first, and then just call Dial with the string-ified IP:port? That'll at least skip DNS.

Before we add new API (e.g. Dialer.DialTCP), we'd really want to understand the problem first.
